Invalidate process object in XPC connection termination watchdog

pvollan opened the pull request:
#### 989daa949491311d8dfb60e86521fe76b726d166
<pre>
Invalidate process object in XPC connection termination watchdog
<a href="https://bugs.webkit.org/show_bug.cgi?id=273773">https://bugs.webkit.org/show_bug.cgi?id=273773</a>
<a href="https://rdar.apple.com/126373957">rdar://126373957</a>

Reviewed by Chris Dumez.

Invalidate process object instead of calling xpc_connection_kill. This will send a termination request
which will terminate the process. This termination approach is only available for WebKit process
extensions.

* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::Client::didRequestProcessTermination):
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
(IPC::Connection::kill):
* Source/WebKit/Platform/cocoa/XPCUtilities.h:
* Source/WebKit/Platform/cocoa/XPCUtilities.mm:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::terminate):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::requestRemoteProcessTermination):
* Source/WebKit/UIProcess/Cocoa/XPCConnectionTerminationWatchdog.h:
* Source/WebKit/UIProcess/Cocoa/XPCConnectionTerminationWatchdog.mm:
(WebKit::XPCConnectionTerminationWatchdog::watchdogTimerFired):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::ProcessLauncher::terminateProcess):
(WebKit::ProcessLauncher::terminateXPCConnection):

Canonical link: <a href="https://commits.webkit.org/279326@main">https://commits.webkit.org/279326@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/888eb9c1a19e2a48423893027ed6387e0a4e70ba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53066 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32403 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5553 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56345 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3789 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39267 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3516 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43038 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2452 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55164 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/30162 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45810 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24166 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/27171 "Found unexpected failure with change (failure)") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1948 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/49023 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3290 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57940 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28207 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3246 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50433 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29427 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46030 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49747 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30346 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7810 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29181 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/Cocoa/XPCConnectionTerminationWatchdog.h:
```diff
@@ -54,6 +54,7 @@ class XPCConnectionTerminationWatchdog {
     OSObjectPtr<xpc_connection_t> m_xpcConnection;
     RunLoop::Timer m_watchdogTimer;
     Ref<ProcessAndUIAssertion> m_assertion;
+    WeakPtr<AuxiliaryProcessProxy> m_process;
```
Pretty sure this needs to be a Ref. Nothing keeps a process alive at this point. So your new code is not likely to run.

cdumez reviewed (changes requested):
It seems this patch fails to fix:
```
AuxiliaryProcessProxy::terminate();
```
which currently relies on IPC::Connection::kill() which no longer works.

cdumez commented on Source/WebKit/UIProcess/Cocoa/XPCConnectionTerminationWatchdog.mm:
```diff
@@ -41,13 +45,19 @@
     
 void XPCConnectionTerminationWatchdog::watchdogTimerFired()
 {
     terminateWithReason(m_xpcConnection.get(), ReasonCode::WatchdogTimerFired, "XPCConnectionTerminationWatchdog::watchdogTimerFired");
```
This doesn't work when using EXTENSIONKIT, so we should only call it in the `#else` case.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -526,6 +526,11 @@ static ASCIILiteral serviceName(const ProcessLauncher::LaunchOptions& launchOpti
 void ProcessLauncher::terminateXPCConnection()
 {
+#if USE(EXTENSIONKIT)
+    if (m_process)
```
Seems the rest of the code using the xpc_connection should likely be called in an `#else` case since I don't believe it does anything when using EXTENSIONKIT.

pvollan replied:
Ah, these a are good catches, I will address in the next patch.

Thanks for reviewing!

cdumez reviewed (commented):
We probably don't even want to build `terminateWithReason()` when using EXTENSIONKIT.

cdumez commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -208,6 +208,9 @@ class Connection : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Connec
         virtual void didClose(Connection&) = 0;
         virtual void didReceiveInvalidMessage(Connection&, MessageName) = 0;
+#if USE(EXTENSIONKIT)
+        virtual void didRequestProcessTermination(bool& wasTerminated) { wasTerminated = false; }
```
Given that we're asking the client to terminate, I think we should use a verb in the name:
e.g. "requestProcessTermination()"

I also think it should return the boolean instead of using an old-style out-parameter.

cdumez replied:
Also, could this be pure virtual? Seems risky for a subclass to NOT override this.

pvollan replied:
That is a good point. I did not add this in the latest patch, since only AuxiliaryProcessProxy is calling `Connection::kill()`, currently. However, please let me know if you still think we should make it pure virtual, I will upload a new PR with that change :)

Thanks for reviewing!

cdumez replied:
Yes, I think it should be pure virtual.. You only provided evidence that it SHOULD be pure virtual so why wouldn't we want it to not be pure virtual?

cdumez commented on Source/WebKit/UIProcess/Cocoa/XPCConnectionTerminationWatchdog.mm:
```diff
@@ -41,13 +45,19 @@
 void XPCConnectionTerminationWatchdog::watchdogTimerFired()
 {
+#if USE(EXTENSIONKIT)
+    if (m_process->extensionProcess())
```
We should store the value returned by `extensionProcess()` in a local to avoid calling the function twice.

cdumez replied:
Should we RELEASE_LOG_ERROR in the else case? Since it would mean we fail to terminate the hung process.

cdumez commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -501,15 +501,22 @@ static ASCIILiteral serviceName(const ProcessLauncher::LaunchOptions& launchOpti
 
-    if (!m_processID)
-        return;
+#if !PLATFORM(COCOA)
```
We're in a Cocoa.mm file so this check doesn't make much sense.. It's essentially dead code after your change.

pvollan replied:
Ah, these are all good point. I will address in the upcoming patch.

Thanks for reviewing!

cdumez commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -208,6 +208,9 @@ class Connection : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Connec
         virtual void didClose(Connection&) = 0;
         virtual void didReceiveInvalidMessage(Connection&, MessageName) = 0;
+#if USE(EXTENSIONKIT)
+        virtual bool requestRemoteProcessTermination() { return false; }
```
Should be pure virtual, as mentioned in my earlier review.

pvollan replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/Cocoa/AuxiliaryProcessProxyCocoa.mm:
```diff
@@ -155,4 +155,16 @@
+        process->invalidate();
+        return true;
+    }
+    RELEASE_LOG_ERROR(Process, "Extension process was not invalidated");
```
Please follow the format of our usual logging. It needs to mentioned the class and method name the logging is called from.

pvollan replied:
Ah, good point, fixed in latest patch.

cdumez commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
```diff
@@ -336,6 +336,7 @@ class GPUConnectionToWebProcess
     void didReceiveInvalidMessage(IPC::Connection&, IPC::MessageName) final;
     void didReceiveMessage(IPC::Connection&, IPC::Decoder&) final;
     bool didReceiveSyncMessage(IPC::Connection&, IPC::Decoder&, UniqueRef<IPC::Encoder>&) final;
+    bool requestRemoteProcessTermination() final { return false; }
```
I thought you told me earlier that there was a single subclass? Now it looks a lot worse since there are many subclasses which basically provide a dummy implementation, which could be defined on the base class instead. This was a step in the wrong direction.

pvollan replied:
Ah, that is a good point, Chris! I am going back to having the dummy implementation in the base class.

Thanks for reviewing!

cdumez commented on Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
```diff
@@ -575,12 +575,16 @@ static void clearNoSenderNotifications(mach_port_t port)
 
 bool Connection::kill()
```
I think likely, a better way to do this is to not build `Connection::kill()` at all when `USE(EXTENSIONKIT)`. I think it is only called in one place: `AuxiliaryProcessProxy::terminate()` so we don't need the new virtual function at all it seems?

pvollan replied:
That is a good point. I wonder if it would be good to keep the virtual function though, since we are also calling `kill()` from `Connection::enqueueIncomingMessage` to terminate a misbehaving child process. What do you think? In the latest PR I kept the virtual function because of this.

cdumez commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -1263,7 +1263,7 @@ void Connection::enqueueIncomingMessage(UniqueRef<Decoder> incomingMessage)
 
         if (isIncomingMessagesThrottlingEnabled() && m_incomingMessages.size() >= maxPendingIncomingMessagesKillingThreshold) {
-            if (kill()) {
+            if (m_client && m_client->requestRemoteProcessTermination()) {
```
Is this called on the main thread?

pvollan replied:
Oh, that is a good point! I will check this and add an assert.

Thanks for reviewing!

cdumez commented on Source/WebKit/UIProcess/Cocoa/AuxiliaryProcessProxyCocoa.mm:
```diff
@@ -155,4 +155,11 @@
 
+void AuxiliaryProcessProxy::requestRemoteProcessTermination()
+{
+    callOnMainRunLoop([protectedThis = Ref { *this }]() {
```
This needs to be dealt with at the call site, not here. AuxiliaryProcessProxy is not a thread-safe class, we shouldn't be calling its member functions on a non-main thread.

pvollan replied:
Ah, good catch, fixed!

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -186,7 +186,7 @@ void AuxiliaryProcessProxy::terminate()
     RELEASE_LOG(Process, "AuxiliaryProcessProxy::terminate: PID=%d", processID());
 
-#if PLATFORM(COCOA)
+#if PLATFORM(COCOA) && !USE(EXTENSIONKIT)
```
I am a little confused why this function doesn't do anything when USE(EXTENSIONKIT). I don't think `processLauncher->terminateProcess();` is enough as the process may not be launching anymore.

pvollan replied:
That is a good point! In the ExtensionKit case, I believe `processLauncher->terminateProcess()` is sufficient, since it will call invalidate on the process object, and thereby terminate the process.

cdumez replied:
I have to think about this. It seems confusing to 100% rely on the process launcher to terminate a process. While it seems OK to involve the process launcher WHILE a process is launching, it is more surprising to do so after it's finished launching.

cdumez replied:
The previous code was trying to kill the process and would only fall back to using the process launcher. The new code doesn't feel as nice, which is unfortunate since it is supposed the new modern code.

pvollan replied:
That is a good point, Chris! I wonder if we should make it more explicit that we hold a reference to the Process launcher throughout the lifetime of the process. Perhaps we could make it a data member instead of having it as a reference data member in AuxiliaryProcessProxy? We can look into this in a follow-up :)

Thanks for reviewing!

cdumez replied:
Right, let's not block fixing the bug. We should land this and we can revisit later if we have something nicer.

pvollan replied:
Yes, that sounds good. Thanks for approving!

pvollan committed:
Invalidate process object in XPC connection termination watchdog
https://bugs.webkit.org/show_bug.cgi?id=273773
rdar://126373957

Reviewed by Chris Dumez.

Invalidate process object instead of calling xpc_connection_kill. This will send a termination request
which will terminate the process. This termination approach is only available for WebKit process
extensions.

* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::Client::didRequestProcessTermination):
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
(IPC::Connection::kill):
* Source/WebKit/Platform/cocoa/XPCUtilities.h:
* Source/WebKit/Platform/cocoa/XPCUtilities.mm:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::terminate):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::requestRemoteProcessTermination):
* Source/WebKit/UIProcess/Cocoa/XPCConnectionTerminationWatchdog.h:
* Source/WebKit/UIProcess/Cocoa/XPCConnectionTerminationWatchdog.mm:
(WebKit::XPCConnectionTerminationWatchdog::watchdogTimerFired):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::ProcessLauncher::terminateProcess):
(WebKit::ProcessLauncher::terminateXPCConnection):

Canonical link: https://commits.webkit.org/279326@main
